Annotate a CSV with plate/well metadata

mcquin opened the pull request:
Resolves #13

Commit:
mock workspace

create backend and metadata subdirectories

Commit:
Implement annotate.

Annotate adds plate and well metadata, as well as external metadata,
to aggregated data and outputs a .CSV file.

Commit:
Add annotate script.

Commit:
import setNames from stats

Commit:
include readr and dplyr namespaces

Commit:
Test option format_broad_cmap

Commit:
Make script executable.

Commit:
add -o/-w flags to command usage

shntnu committed:
Cosmetic changes based on suggestions by lint.

shntnu committed:
Added rlang dependency so we can use .data

shntnu committed:
Use .data to overcome "no global bindings warnings" (shows up when CHECK is run on package).

shntnu committed:
Cosmetic changes based on lint

shntnu commented:
I used `.data` to overcome the "no visible bindings issue"
https://stackoverflow.com/questions/9439256/how-can-i-handle-r-cmd-check-no-visible-binding-for-global-variable-notes-when
I previously had used `rlang::sym` to overcome this
https://github.com/cytomining/cytominer/blob/649ba287f7f8ff7faa0992fba276ebee1d22354f/R/drop_na_columns.R#L35-L39
but changed my mind after reading 

https://community.rstudio.com/t/should-tidyeval-be-abandoned/
especially
https://community.rstudio.com/t/should-tidyeval-be-abandoned/2238/5

I find tidyeval very helpful but it does cause some headaches. I anticipate this getting easier over the next year once it matures.

shntnu committed:
Fixed typo

shntnu committed:
Added docs

mcquin commented:
👏 @shntnu!

Commit:
describe format_broad_cmap test

shntnu commented on R/annotate.R:
```diff
@@ -0,0 +1,202 @@
+#' @param batch_id            Batch ID.
+#' @param plate_id            Plate ID.
+#' @param cell_id             Optional cell ID. default: \code{NULL}.
+#' @param external_metadata   Optional external metadata to join with, as a JSON file. default: \code{NULL}.
```
as a CSV file, not JSON

Commit:
Test external_metadata

* Update documentation:
  - Requires CSV file
  - Requires Metadata_Plate and Metadata_Well columns
* Fail fast if required columns aren't present

shntnu reviewed (commented):
@mcquin thanks for adding that test!

shntnu commented on tests/testthat/test-annotate.R:
```diff
@@ -0,0 +1,117 @@
+
+  external_metadata_csv <- tempfile("external_m")
+
+  external_metadata <- tibble(
```
`dplyr::tibble`

shntnu commented on tests/testthat/test-annotate.R:
```diff
@@ -0,0 +1,117 @@
+
+  external_metadata_csv <- tempfile("external_m")
+
+  external_metadata <- tibble(
```
`dplyr::tibble`

shntnu reviewed (changes requested):
Looks good except for two changes suggested. Awesome that we have a test for external metadata!

Commit:
specify dplyr namespace

shntnu committed:
Fix lists
(see https://cran.r-project.org/web/packages/roxygen2/vignettes/formatting.html)

shntnu reviewed (commented):
Minor fix for docs, looks good @mcquin !
